src/client/particles.h

```diff
@@ -136,44 +138,48 @@ class Particle : public scene::ISceneNode
 	bool m_collision_removal;
 	bool m_object_collision;
 	bool m_vertical;
-	v3s16 m_camera_offset;
 	struct TileAnimationParams m_animation;
 	float m_animation_time = 0.0f;
 	int m_animation_frame = 0;
 	u8 m_glow;
 	float m_alpha = 0.0f;
+
+	ParticleSpawner *m_parent = nullptr;
+	// Used if not spawned from a particlespawner
+	std::unique_ptr<ClientParticleTexture> m_owned_texture;
 };
 
 class ParticleSpawner
 {
 public:
 	ParticleSpawner(IGameDef *gamedef,
 		LocalPlayer *player,
-		const ParticleSpawnerParameters &p,
+		const ParticleSpawnerParameters &params,
 		u16 attached_id,
-		std::unique_ptr<ClientTexture[]> &texpool,
-		size_t texcount,
-		ParticleManager* p_manager);
+		std::vector<ClientParticleTexture> &&texpool,
+		ParticleManager *p_manager);
 
 	void step(float dtime, ClientEnvironment *env);
 
-	size_t m_active;
-
 	bool getExpired() const
 	{ return m_dying || (p.amount <= 0 && p.time != 0); }
 	void setDying() { m_dying = true; }
 
+	bool hasActive() const { return m_active != 0; }
+	void decrActive() { m_active -= 1; }
```

Comment:
Is there no list where this condition could be checked directly rather than risking these values to go out-of-sync?

Reply:
No, a particlespawner has no container for its particles / its children.
The particles live in the manager's `m_particles`.